spring-kafka/src/test/java/org/springframework/kafka/retrytopic/AsyncCompletableFutureRetryTopicScenarioTests.java

```diff
@@ -0,0 +1,1263 @@
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.TestPropertySource;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+/**
+ * @author Sanghyeok An
+ * @since 3.3.0
+ */
+
+@SpringJUnitConfig
+@DirtiesContext
+@EmbeddedKafka
+@TestPropertySource(properties = { "five.attempts=5", "kafka.template=customKafkaTemplate"})
+public class AsyncCompletableFutureRetryTopicScenarioTests {
+
+	private final static String MAIN_TOPIC_CONTAINER_FACTORY = "kafkaListenerContainerFactory";
+
+	public final static String TEST_TOPIC0 = "myRetryTopic0";
+
+	public final static String TEST_TOPIC1 = "myRetryTopic1";
+
+	public final static String TEST_TOPIC2 = "myRetryTopic2";
+
+	public final static String TEST_TOPIC3 = "myRetryTopic3";
+
+	public final static String TEST_TOPIC4 = "myRetryTopic4";
+
+	public final static String TEST_TOPIC5 = "myRetryTopic5";
+
+	public final static String TEST_TOPIC6 = "myRetryTopic6";
+
+	@Autowired
+	private KafkaTemplate<String, String> kafkaTemplate;
+
+	@Autowired
+	private CountDownLatchContainer latchContainer;
+
+	@Autowired
+	DestinationTopicContainer topicContainer;
+
+	@Test
+	void allFailCaseTest(
+			@Autowired TestTopicListener0 zeroTopicListener,
+			@Autowired MyCustomDltProcessor myCustomDltProcessor0,
+			@Autowired ThreadPoolTaskExecutor executor) {
+		// All Fail case.
+		String shortFailedMsg1 = "0";
+		String shortFailedMsg2 = "1";
+		String shortFailedMsg3 = "2";
+		DestinationTopic destinationTopic = topicContainer.getNextDestinationTopicFor("0-topicId", TEST_TOPIC0);
+
+		String expectedRetryTopic = TEST_TOPIC0 + "-retry";
+		String[] expectedReceivedMsgs = {
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3,
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3
+		};
+		String[] expectedReceivedTopics = {
+				TEST_TOPIC0,
+				TEST_TOPIC0,
+				TEST_TOPIC0,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic,
+				expectedRetryTopic
+		};
+		String[] expectedDltMsgs = {
+				shortFailedMsg1,
+				shortFailedMsg2,
+				shortFailedMsg3
+		};
+
+		// When
+		kafkaTemplate.send(TEST_TOPIC0, shortFailedMsg1);
+		kafkaTemplate.send(TEST_TOPIC0, shortFailedMsg2);
+		kafkaTemplate.send(TEST_TOPIC0, shortFailedMsg3);
+
+		// Then
+		assertThat(awaitLatch(latchContainer.countDownLatch0)).isTrue();
+		assertThat(awaitLatch(latchContainer.dltCountdownLatch0)).isTrue();
+
+		waitAWhile(executor, 10000);
```

Comment:
I removed `waitAWhile(executor, 10000)` from all scenario tests! 

> what is the difference (at a glance) between those integration tests and these scenarios, please?

@artembilan My bad, i missed it. 🙇‍♂️ 
1. The `RetryTopicClassLevelIntegrationTests` class is base class of `Async(CompletableFuture | Mono)RetryTopicClassLevelIntegrationTests`. 
2. These async retry integration test cases follow the rules of the `RetryTopicClassLevelIntegrationTests`. In other words, they demonstrate that the async retry feature works well in the synchronous conditions. 
3. However, in the async condition, the message with offset 1 can fail earlier than the message with offset 0. this situation never occur in the synchronous condition. because, `ConsumerRecord` is invoked one by one.
4. The async scenario tests demonstrate that the async retry feature works well with the situation which i mentioned at 3.

I am concerned that async retry cases may affect the `offset` because the async retry feature follows the rules of `FailedRecordTracker` and `SeekUtils`. As a result of seeking the wrong offset, i am worried about the cases where a message coulde be re-consumed infinitely or some offsets could be skipped to be retried.

the async scenario test cases demonstrate this.

Reply:
I think the async retry feature should include not only `CompletableFuture<?>` and `Mono<?>` test cases  based on `RetryTopicClassLevelIntegrationTests` but also `Async*ScenarioTests`. 

`CompletableFuture<?>` and `Mono<?>` test cases  based on `RetryTopicClassLevelIntegrationTests` demonstrate sequential failures (e.g., offset 0 fail -> offset 1 fail -> ....). 

However, `Async*ScenarioTests` demonstrate non-sequential failures. 

For example, i imagine the following situations.
1. put `Mono<?>` instance with Offset 0 to Executor.
2. put `Mono<?>` instance with Offset 1 to Executor. 
3. put `Mono<?>` instance with Offset 2 to Executor. 
4. `Mono<?>` instance with Offset 2 succeed after 100ms.
5. `Mono<?>` instance with Offset 0 failed after 500ms.
6. `Mono<?>` instance with Offset 0 should be retry. (Don't retry offset 1, 2)
7. `Mono<?>` instance with Offset 1 failed after 600ms.
8. `Mono<?>` instance with Offset 1 should be retry. (Don't retry offset 0, 2)


```java
kafkaTemplate.send(TEST_TOPIC3, TestTopicListener3.LONG_FAIL_MSG);
kafkaTemplate.send(TEST_TOPIC3, TestTopicListener3.LONG_FAIL_MSG);
kafkaTemplate.send(TEST_TOPIC3, TestTopicListener3.SHORT_SUCCESS_MSG);
kafkaTemplate.send(TEST_TOPIC3, TestTopicListener3.SHORT_SUCCESS_MSG);
kafkaTemplate.send(TEST_TOPIC3, TestTopicListener3.SHORT_SUCCESS_MSG);
```